File: src/app/src/main/java/com/huawei/touchmenot/java/hms/hand/Hand.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.touchmenot.java.hms.hand;

import android.util.SparseArray;

import androidx.core.util.SparseArrayKt;

import com.huawei.hms.mlsdk.common.MLAnalyzer;
import com.huawei.hms.mlsdk.handkeypoint.MLHandKeypoint;
import com.huawei.hms.mlsdk.handkeypoint.MLHandKeypoints;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import kotlin.collections.CollectionsKt;
import kotlin.jvm.internal.Intrinsics;

public class Hand {

    private ArrayList<MLHandKeypoint> firstFinger = new ArrayList<>();
    private ArrayList<MLHandKeypoint> middleFinger = new ArrayList<>();
    private ArrayList<MLHandKeypoint> ringFinger = new ArrayList<>();
    private ArrayList<MLHandKeypoint> littleFinger = new ArrayList<>();
    private ArrayList<MLHandKeypoint> thumb = new ArrayList<>();

    private MLHandKeypoint wrist;
    public static int ACCEPTED_RANGE_IN_DEGREE = 30;
    private static final int DEGREE = 360;
    private static final int FINGER_COUNT_LESS_THAN = 6;

    public static String analyzeHandsAndGetNumber(MLAnalyzer.Result<MLHandKeypoints> result) {
        ArrayList hands = new ArrayList();
        int number = 0;
        if (result != null) {
            SparseArray analyseList = result.getAnalyseList();
            Intrinsics.checkNotNullExpressionValue(analyseList, "result.analyseList");
            if (analyseList != null) {
                Iterator keyIterator = (Iterator) SparseArrayKt.keyIterator(analyseList);
                while (keyIterator.hasNext()) {
                    int key = ((Number) keyIterator.next()).intValue();
                    hands.add(new Hand());
                    analyseList = result.getAnalyseList();
                    Intrinsics.checkNotNullExpressionValue(analyseList, "result.analyseList");
                    Iterator iterator = SparseArrayKt.valueIterator(analyseList);
                    Hand hand;
                    List list;
                    for (Iterator var11 = iterator; var11.hasNext(); number += hand.createHand(list).getNumber()) {
                        MLHandKeypoints value = (MLHandKeypoints) var11.next();
                        hand = (Hand) CollectionsKt.last((List) hands);
                        Intrinsics.checkNotNullExpressionValue(value, "value");
                        list = value.getHandKeypoints();
                        Intrinsics.checkNotNullExpressionValue(list, "value.handKeypoints");
                    }
                }
            }

        }
        if (number < FINGER_COUNT_LESS_THAN)
            return String.valueOf(number);
        else
            return "";
    }

    private Hand createHand(List<MLHandKeypoint> keyPoints) {
        for (MLHandKeypoint keyPoint : keyPoints) {
            switch (keyPoint.getType()) {
                case MLHandKeypoint.TYPE_WRIST:
                    wrist = keyPoint;
                    break;

                case MLHandKeypoint.TYPE_THUMB_FIRST:
                    thumb.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_THUMB_SECOND:
                    thumb.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_THUMB_THIRD:
                    thumb.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_THUMB_FOURTH:
                    thumb.add(keyPoint);
                    break;

                case MLHandKeypoint.TYPE_FOREFINGER_FIRST:
                    firstFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_FOREFINGER_SECOND:
                    firstFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_FOREFINGER_THIRD:
                    firstFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_FOREFINGER_FOURTH:
                    firstFinger.add(keyPoint);
                    break;

                case MLHandKeypoint.TYPE_MIDDLE_FINGER_FIRST:
                    middleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_MIDDLE_FINGER_SECOND:
                    middleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_MIDDLE_FINGER_THIRD:
                    middleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_MIDDLE_FINGER_FOURTH:
                    middleFinger.add(keyPoint);
                    break;

                case MLHandKeypoint.TYPE_RING_FINGER_FIRST:
                    ringFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_RING_FINGER_SECOND:
                    ringFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_RING_FINGER_THIRD:
                    ringFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_RING_FINGER_FOURTH:
                    ringFinger.add(keyPoint);
                    break;

                case MLHandKeypoint.TYPE_LITTLE_FINGER_FIRST:
                    littleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_LITTLE_FINGER_SECOND:
                    littleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_LITTLE_FINGER_THIRD:
                    littleFinger.add(keyPoint);
                    break;
                case MLHandKeypoint.TYPE_LITTLE_FINGER_FOURTH:
                    littleFinger.add(keyPoint);
                    break;
            }
        }
        return this;
    }

    private int getNumber() {
        int number = 0;
        if (wrist != null) {
            ArrayList var10001 = this.firstFinger;
            MLHandKeypoint var10002 = this.wrist;
            Intrinsics.checkNotNull(var10002);
            if (this.isFingerUp(var10001, var10002)) {
                ++number;
            }

            var10001 = this.middleFinger;
            var10002 = this.wrist;
            Intrinsics.checkNotNull(var10002);
            if (this.isFingerUp(var10001, var10002)) {
                ++number;
            }

            var10001 = this.ringFinger;
            var10002 = this.wrist;
            Intrinsics.checkNotNull(var10002);
            if (this.isFingerUp(var10001, var10002)) {
                ++number;
            }

            var10001 = this.littleFinger;
            var10002 = this.wrist;
            Intrinsics.checkNotNull(var10002);
            if (this.isFingerUp(var10001, var10002)) {
                ++number;
            }

            var10001 = this.thumb;
            var10002 = this.wrist;
            Intrinsics.checkNotNull(var10002);
            if (this.isFingerUp(var10001, var10002)) {
                ++number;
            }
        }
        return number;
    }

    private boolean isFingerUp(ArrayList<MLHandKeypoint> points, MLHandKeypoint wrist) {
        if (points.size() == 4) {
            double degreeFirst2Last = this.getDegree((MLHandKeypoint) CollectionsKt.first((List) points), (MLHandKeypoint) CollectionsKt.last((List) points));
            MLHandKeypoint var10001 = (MLHandKeypoint) CollectionsKt.first((List) points);
            Object var10002 = points.get(1);
            Intrinsics.checkNotNullExpressionValue(var10002, "points[1]");
            double degreeFirst2Second = this.getDegree(var10001, (MLHandKeypoint) var10002);
            double var7 = degreeFirst2Last - degreeFirst2Second;
            boolean var9 = false;
            if (Math.abs(var7) < (double) ACCEPTED_RANGE_IN_DEGREE) {
                return true;
            }
        }
        return false;
    }

    private double getDegree(MLHandKeypoint point1, MLHandKeypoint point2) {
        float width = point2.getPointX() - point1.getPointX();
        float height = point2.getPointY() - point1.getPointY();
        double angle = height;
        double val = width;
        boolean value = false;
        double theta = Math.atan2(angle, val);
        theta++;
        angle = Math.toDegrees(theta);
        if (angle < 0) {
            angle += (double) DEGREE;
        }
        return angle;
    }

}

File: src/app/src/main/java/com/huawei/touchmenot/kotlin/hms/hand/Hand.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.huawei.touchmenot.kotlin.hms.hand

import androidx.core.util.keyIterator
import androidx.core.util.valueIterator
import com.huawei.hms.mlsdk.common.MLAnalyzer
import com.huawei.hms.mlsdk.handkeypoint.MLHandKeypoint
import com.huawei.hms.mlsdk.handkeypoint.MLHandKeypoints
import kotlin.math.abs
import kotlin.math.atan2

class Hand {

    companion object {
        private const val ACCEPTED_RANGE_IN_DEGREE = 30
        private const val DEGREE = 360
        private const val FINGER_COUNT_LESS_THAN = 6

        /**
         *  Creates hands and checks if fingers are up or not. Counts fingers that are up and returns as
         *  String.
         */
        fun analyzeHandsAndGetNumber(result: MLAnalyzer.Result<MLHandKeypoints>): String {
            val hands = ArrayList<Hand>()
            var number = 0

            for (key in result.analyseList.keyIterator()) {
                hands.add(Hand())

                for (value in result.analyseList.valueIterator()) {
                    number += hands.last().createHand(value.handKeypoints).getNumber()
                }
            }
            if (number < FINGER_COUNT_LESS_THAN)
                return number.toString()
            else
                return ""
        }
    }

    var firstFinger = arrayListOf<MLHandKeypoint>()
    var middleFinger = arrayListOf<MLHandKeypoint>()
    var ringFinger = arrayListOf<MLHandKeypoint>()
    var littleFinger = arrayListOf<MLHandKeypoint>()
    var thumb = arrayListOf<MLHandKeypoint>()
    var wrist: MLHandKeypoint? = null

    /**
     *  Creates and returns a hand with first finger, middle finger, ring finger, little finger,
     *  thumb and wrist.
     */
    fun createHand(keyPoints: MutableList<MLHandKeypoint>): Hand {

        for (keyPoint in keyPoints) {

            when (keyPoint.type) {
                MLHandKeypoint.TYPE_FOREFINGER_FIRST -> firstFinger.add(keyPoint)
                MLHandKeypoint.TYPE_FOREFINGER_SECOND -> firstFinger.add(keyPoint)
                MLHandKeypoint.TYPE_FOREFINGER_THIRD -> firstFinger.add(keyPoint)
                MLHandKeypoint.TYPE_FOREFINGER_FOURTH -> firstFinger.add(keyPoint)

                MLHandKeypoint.TYPE_MIDDLE_FINGER_FIRST -> middleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_MIDDLE_FINGER_SECOND -> middleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_MIDDLE_FINGER_THIRD -> middleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_MIDDLE_FINGER_FOURTH -> middleFinger.add(keyPoint)

                MLHandKeypoint.TYPE_RING_FINGER_FIRST -> ringFinger.add(keyPoint)
                MLHandKeypoint.TYPE_RING_FINGER_SECOND -> ringFinger.add(keyPoint)
                MLHandKeypoint.TYPE_RING_FINGER_THIRD -> ringFinger.add(keyPoint)
                MLHandKeypoint.TYPE_RING_FINGER_FOURTH -> ringFinger.add(keyPoint)

                MLHandKeypoint.TYPE_LITTLE_FINGER_FIRST -> littleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_LITTLE_FINGER_SECOND -> littleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_LITTLE_FINGER_THIRD -> littleFinger.add(keyPoint)
                MLHandKeypoint.TYPE_LITTLE_FINGER_FOURTH -> littleFinger.add(keyPoint)

                MLHandKeypoint.TYPE_THUMB_FIRST -> thumb.add(keyPoint)
                MLHandKeypoint.TYPE_THUMB_SECOND -> thumb.add(keyPoint)
                MLHandKeypoint.TYPE_THUMB_THIRD -> thumb.add(keyPoint)
                MLHandKeypoint.TYPE_THUMB_FOURTH -> thumb.add(keyPoint)

                MLHandKeypoint.TYPE_WRIST -> wrist = keyPoint
            }
        }

        return this
    }

    /**
     *  Detects the number shown by hand. Counts fingers that are up.
     */
    fun getNumber(): Int {
        var number = 0

        if (wrist != null) {

            if (isFingerUp(firstFinger, wrist!!))
                number++

            if (isFingerUp(middleFinger, wrist!!))
                number++

            if (isFingerUp(ringFinger, wrist!!))
                number++

            if (isFingerUp(littleFinger, wrist!!))
                number++

            if (isFingerUp(thumb, wrist!!))
                number++
        }

        return number
    }

    /**
     *  A finger is considered up if the difference between, degree calculated between
     *  first and last points of the finger and degree between first and second points of the finger
     *  is in accepted range.
     */
    private fun isFingerUp(points: ArrayList<MLHandKeypoint>, wrist: MLHandKeypoint): Boolean {

        if (points.size == 4) {

            val degreeFirst2Last = getDegree(points.first(), points.last())
            val degreeFirst2Second = getDegree(points.first(), points[1])

            if (abs(degreeFirst2Last - degreeFirst2Second) < ACCEPTED_RANGE_IN_DEGREE)
                return true
        }

        return false
    }

    /**
     *  Finds the degree between to points.
     */
    private fun getDegree(point1: MLHandKeypoint, point2: MLHandKeypoint): Double {

        val width = point2.pointX - point1.pointX
        val height = point2.pointY - point1.pointY

        //Returns the angle theta from the conversion of rectangular coordinates (x, y) to polar coordinates (r, theta)
        var theta = atan2(height.toDouble(), width.toDouble())

        //Rotate theta angle PI/2 radian clockwise.
        theta += Math.PI / 2.0

        //Convert radian to degree.
        var angle = Math.toDegrees(theta)

        if (angle < 0) {
            angle += DEGREE
        }
        return angle
    }

    private fun analyzeHandsAndGetNumber(result: MLAnalyzer.Result<MLHandKeypoints>): String {
        val hands = ArrayList<Hand>()
        var number = 0

        for (key in result.analyseList.keyIterator()) {
            hands.add(Hand())

            for (value in result.analyseList.valueIterator()) {
                number += hands.last().createHand(value.handKeypoints).getNumber()
            }
        }

        return number.toString()
    }

}
